autofaiss/external/distributed.py

```diff
@@ -0,0 +1,367 @@
+
+
+def _yield_embeddings_batch(
+    embeddings_paths: List[str],
+    chunk_sizes: List[int],
+    file_system: fsspec.AbstractFileSystem,
+    start: int,
+    end: int,
+    embedding_column_name: str,
+    file_format: str,
+    id_columns: Optional[List[str]] = None,
+):
+    """Lazy load a batch of embeddings."""
+    if isinstance(file_system, PyArrowHDFS):
+        file_system = get_filesystem_class("hdfs")()
+    cur_start = cur_end = 0
+    for chunk_size, file_path in zip(chunk_sizes, embeddings_paths):
+        cur_end += chunk_size
+        if cur_end < start:
+            cur_start += chunk_size
+            continue
+        slice_start = max(0, start - cur_start)
+        slice_end = min(chunk_size, end - cur_start)
+        with get_matrix_reader(file_format, file_system, file_path, embedding_column_name, id_columns) as matrix_reader:
+            yield matrix_reader.get_lazy_array().get_rows(start=slice_start, end=slice_end)
+        if cur_end > end:
+            break
+        cur_start += chunk_size
+
+
+def _get_index_from_bytes(index_bytes: Union[bytearray, bytes]) -> faiss.Index:
+    """Transforms a bytearray containing a faiss index into the corresponding object."""
+
+    with NamedTemporaryFile(delete=False) as output_file:
+        output_file.write(index_bytes)
+        tmp_name = output_file.name
+
+    b = faiss.read_index(tmp_name)
+    os.remove(tmp_name)
+    return b
+
+
+def _get_bytes_from_index(index: faiss.Index) -> bytearray:
+    """Transforms a faiss index into a bytearray."""
+
+    with NamedTemporaryFile(delete=False) as output_file:
+        faiss.write_index(index, output_file.name)
+        tmp_name = output_file.name
+
+    with open(tmp_name, "rb") as index_file:
+        b = index_file.read()
+        os.remove(tmp_name)
+        return bytearray(b)
+
+
+def _generate_small_index_file_name(batch_id: int) -> str:
+    return f"index_{batch_id}"
+
+
+def _save_small_index(index: faiss.Index, batch_id: int, small_indices_folder: str):
+    """Save index for one batch."""
+    fs = _get_file_system(small_indices_folder)
+    fs.mkdirs(small_indices_folder, exist_ok=True)
+    small_index_filename = _generate_small_index_file_name(batch_id)
+    with fsspec.open(small_index_filename, "wb").open() as f:
+        faiss.write_index(index, faiss.PyCallbackIOWriter(f.write))
+    dest_filepath = os.path.join(small_indices_folder, small_index_filename)
+    fs.put(small_index_filename, dest_filepath)
+
+
+def _add_index(
+    start: int,
+    end: int,
+    broadcast_trained_index_bytes,
+    file_system: fsspec.AbstractFileSystem,
+    chunk_sizes: List[int],
+    embeddings_file_paths: List[str],
+    embedding_column_name: str,
+    batch_id: int,
+    small_indices_folder: str,
+    file_format: str,
+    id_columns: Optional[List[str]] = None,
+    num_cores: Optional[int] = None,
+    embedding_ids_df_handler: Optional[Callable[[pd.DataFrame, int], Any]] = None,
+):
```

Comment:
This one is hard to give a return type, it could be a customized function that passed from build.py. For now, I reuse the same typing as that in the create_index.